addons/website_sale_coupon/models/sale_order.py

```diff
@@ -93,3 +113,29 @@ def _gc_abandoned_coupons(self, *args, **kwargs):
             coupon.sales_order_id.applied_coupon_ids -= coupon
         coupon_to_reset.write({'state': 'new'})
         coupon_to_reset.mapped('sales_order_id').recompute_coupon_lines()
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    def unlink(self):
+        reward_lines = self.filtered(lambda line: line.is_reward_line)
+        status = request.session.get('pending_coupon') or {}
+        code = status.get('code')
+
+        if request and reward_lines and code:
```

Comment:
`if request` is generally a red flag that something is not written the best it could

In this case, you should just do this in `update_json` route (check which route is called when you remove a product in your debugger) and not in a unlink overide, which is called only on purpose by the user.

Reply:
I modeled it after the coupon reactivation in `sale.order.line` unlink method in `sale_coupon`. I do agree though we should separate concerns wherever possible.